components/ws-proxy/pkg/sshproxy/server.go

```diff
@@ -106,44 +107,57 @@ func New(signers []ssh.Signer, workspaceInfoProvider p.WorkspaceInfoProvider, he
 		ServerVersion: "SSH-2.0-GITPOD-GATEWAY",
 		PasswordCallback: func(conn ssh.ConnMetadata, password []byte) (perm *ssh.Permissions, err error) {
 			workspaceId, ownerToken := conn.User(), string(password)
-			wsInfo, err := server.Authenticator(workspaceId, ownerToken)
+			wsInfo, err := server.GetWorkspaceInfo(workspaceId)
 			defer func() {
 				server.TrackSSHConnection(wsInfo, "auth", err)
 			}()
 			if err != nil {
-				if err == ErrAuthFailed {
-					return
-				}
 				args := strings.Split(conn.User(), "#")
 				if len(args) != 2 {
 					return
 				}
 				workspaceId, ownerToken = args[0], args[1]
-				wsInfo, err = server.Authenticator(workspaceId, ownerToken)
-				if err == nil {
-					err = ErrMissPrivateKey
+				wsInfo, err := server.GetWorkspaceInfo(workspaceId)
+				if err != nil {
+					return nil, ErrWorkspaceNotFound
+				}
+				if wsInfo.Auth.OwnerToken == ownerToken {
+					return nil, ErrMissPrivateKey
 				}
-				return
+				return nil, ErrAuthFailed
+			}
+			if wsInfo.Auth.OwnerToken != ownerToken {
+				return nil, ErrAuthFailed
 			}
 			return &ssh.Permissions{
 				Extensions: map[string]string{
 					"workspaceId": workspaceId,
 				},
 			}, nil
 		},
-		PublicKeyCallback: func(conn ssh.ConnMetadata, key ssh.PublicKey) (*ssh.Permissions, error) {
+		PublicKeyCallback: func(conn ssh.ConnMetadata, pk ssh.PublicKey) (*ssh.Permissions, error) {
 			args := strings.Split(conn.User(), "#")
-			// workspaceId#ownerToken
-			if len(args) != 2 {
-				return nil, ErrUsernameFormat
+			workspaceId := args[0]
+			wsInfo, err := server.GetWorkspaceInfo(workspaceId)
+			if err != nil {
+				return nil, err
 			}
-			workspaceId, ownerToken := args[0], args[1]
-			wsInfo, err := server.Authenticator(workspaceId, ownerToken)
 			defer func() {
 				server.TrackSSHConnection(wsInfo, "auth", err)
 			}()
-			if err != nil {
-				return nil, err
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+			defer cancel()
+			ok, _ := server.VerifyPublicKey(ctx, wsInfo, pk)
+			if ok {
+				return &ssh.Permissions{
+					Extensions: map[string]string{
+						"workspaceId": workspaceId,
+					},
+				}, nil
+			}
+			// workspaceId#ownerToken
+			if len(args) != 2 || wsInfo.Auth.OwnerToken != args[1] {
+				return nil, ErrAuthFailed
```

Comment:
the same here, should not we assign  ErrAuthFailed to err first to capture in analytics?